group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -0,0 +1,454 @@
+                        .setPartitionIndex(partition.partitionIndex())
+                        .setErrorCode(Errors.NONE.code()));
+
+                    final OptionalInt leaderEpoch = partition.committedLeaderEpoch() == RecordBatch.NO_PARTITION_LEADER_EPOCH ?
+                        OptionalInt.empty() : OptionalInt.of(partition.committedLeaderEpoch());
+                    final String metadata = partition.committedMetadata() == null ?
+                        OffsetAndMetadata.NO_METADATA : partition.committedMetadata();
+                    final long commitTimestampMs = partition.commitTimestamp() == OffsetCommitRequest.DEFAULT_TIMESTAMP ?
+                        currentTimeMs : partition.commitTimestamp();
+
+                    final OffsetAndMetadata offsetAndMetadata = new OffsetAndMetadata(
+                        partition.committedOffset(),
+                        leaderEpoch,
+                        metadata,
+                        commitTimestampMs,
+                        expireTimestampMs
+                    );
+
+                    records.add(RecordHelpers.newOffsetCommitRecord(
+                        request.groupId(),
+                        topic.name(),
+                        partition.partitionIndex(),
+                        offsetAndMetadata,
+                        metadataImage.features().metadataVersion()
+                    ));
+                }
+            });
+        });
+
+        return new CoordinatorResult<>(records, response);
+    }
+
+    /**
+     * Replays OffsetCommitKey/Value to update or delete the corresponding offsets.
+     *
+     * @param key   A OffsetCommitKey key.
+     * @param value A OffsetCommitValue value.
+     */
+    public void replay(
+        OffsetCommitKey key,
+        OffsetCommitValue value
+    ) {
+        final String groupId = key.group();
+        final TopicPartition tp = new TopicPartition(key.topic(), key.partition());
+
+        if (value != null) {
+            // Ensures that there is a corresponding group for the offsets. If a group does
+            // not exist, a generic group is created as a so-called "simple group".
+            try {
+                groupMetadataManager.group(groupId);
+            } catch (GroupIdNotFoundException ex) {
+                groupMetadataManager.getOrMaybeCreateGenericGroup(groupId, true);
+            }
+
+            final OffsetAndMetadata offsetAndMetadata = OffsetAndMetadata.fromRecord(value);
+            TimelineHashMap<TopicPartition, OffsetAndMetadata> offsets = offsetsByGroup.get(groupId);
+            if (offsets == null) {
+                offsets = new TimelineHashMap<>(snapshotRegistry, 0);
+                offsetsByGroup.put(groupId, offsets);
+            }
+
+            offsets.put(tp, offsetAndMetadata);
+        } else {
+            TimelineHashMap<TopicPartition, OffsetAndMetadata> offsets = offsetsByGroup.get(groupId);
```

Comment:
nit: should we add `// Tombstone.`?

Reply:
This is not necessary and we did not do this for all the other replay methods.

Reply:
i did it for GroupMetadata record 😅 i'll remove it.